### PR TITLE
(WIP) Introduce new arena system

### PR DESCRIPTION
**Pull request by hskim881028**
- [아레나 개선 기획 문서](https://www.notion.so/planetarium/0509-39d32d8767b246418f638367f19b9cd9)
---
- Create `JoinArena` action 
- Create `BattleArena` action (work in progress)
   - This is the action that replaces `RankingBattle`
- Create `ArenaAvatar` state
   - This state has the avatar information you need in the arena. (equipment, costumes, avatar level)
- Create `ArenaParticipants` 
   - ArenaParticipants has a list of participants for the arena
- Create `ArenaScore`
    -  ArenaScore only has a score.
- Create `ArenaInformation` 
   - ArenaInformation has win, lose, ticket
 - Create  `Arena` Sheet
   - This sheet contains arena related information
  ---
- Testbed 9c branch : [feature/new-arena-testbed](https://github.com/planetarium/NineChronicles/tree/feature/new-arena-testbed)
   - you can test `JoinArena` Action 
   - you can see `ArenaState` and `ArenaAvatarState`
   
image

### CODE REVIEW

File: Lib9c/Action/ArenaBattle.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+            var questListAddress = avatarAddress.Derive(LegacyQuestListKey);
+
+            if (context.Rehearsal)
+            {
+                return states
+                    .SetState(arenaStateAddress, MarkChanged)
+                    .SetState(arenaAvatarStateAddress, MarkChanged)
+                    .SetState(avatarAddress, MarkChanged)
+                    .SetState(inventoryAddress, MarkChanged)
+                    .SetState(questListAddress, MarkChanged);
```
Comment (boscohyun) [inline Lib9c/Action/ArenaBattle.cs:65]: We do not need to mark some addresses now. Please return the states as it is when rehearsal.

```suggestion
                return states;
```

File: Lib9c/Action/ArenaBattle.cs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+using Nekoyume.Extensions;
+using Nekoyume.Helper;
+using Nekoyume.Model.Item;
+using Nekoyume.Model.State;
+using Nekoyume.TableData;
+using static Lib9c.SerializeKeys;
+
+namespace Nekoyume.Action
+{
+    [Serializable]
```
Comment (boscohyun) [inline Lib9c/Action/ArenaBattle.cs:18]: A summary is needed to list the history of action changes.

```suggestion
{
    /// <summary>
    /// Introduced at: https://github.com/planetarium/lib9c/pull/1006
    /// </summary>
    [Serializable]
```

File: Lib9c/Action/JoinArena.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+using Bencodex.Types;
+using Libplanet;
+using Libplanet.Action;
+using Nekoyume.Helper;
+using Nekoyume.Model.State;
+using Nekoyume.TableData;
+
+namespace Nekoyume.Action
+{
+    [Serializable]
```
Comment (boscohyun) [inline Lib9c/Action/JoinArena.cs:14]:
```suggestion
    /// <summary>
    /// Introduced at https://github.com/planetarium/lib9c/pull/1006
    /// </summary>
    [Serializable]
```

File: Lib9c/Action/JoinArena.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            var states = context.PreviousStates;
+            var currentBlockIndex = context.BlockIndex;
+            var arenaAddress = Addresses.Arena;
+
+            var arenaAvatarStateAddress = ArenaAvatarState.DeriveAddress(avatarAddress);
+            if (context.Rehearsal)
+            {
+                return states.SetState(ArenaState.DeriveAddress(currentBlockIndex), MarkChanged)
+                    .SetState(arenaAvatarStateAddress, MarkChanged)
+                    .MarkBalanceChanged(GoldCurrencyMock, context.Signer, arenaAddress);
```
Comment (boscohyun) [inline Lib9c/Action/JoinArena.cs:51]:
```suggestion
                return states;
```

File: Lib9c/Action/NotEnoughWinException.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Runtime.Serialization;
+
+namespace Nekoyume.Action
+{
+    public class NotEnoughWinException : Exception
```
Comment (boscohyun) [inline Lib9c/Action/NotEnoughWinException.cs:6]:
```suggestion
    [Serializable]
    public class NotEnoughWinException : Exception
```

File: Lib9c/Model/State/AlreadyContainsException.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+using System;
+using System.Runtime.Serialization;
+
+namespace Nekoyume.Model.State
+{
+    public class AlreadyContainsException : Exception
```
Comment (boscohyun) [inline Lib9c/Model/State/AlreadyContainsException.cs:6]:
```suggestion
    [Serializable]
    public class AlreadyContainsException : Exception
```

File: Lib9c/Model/State/NotExistException.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
+using System;
+using System.Runtime.Serialization;
+
+namespace Nekoyume.Model.State
+{
+    public class NotExistException : Exception
```
Comment (boscohyun) [inline Lib9c/Model/State/NotExistException.cs:6]:
```suggestion
    [Serializable]
    public class NotExistException : Exception
```

**Review by boscohyun — COMMENTED** [review]
New exceptions are needed to testing in [here](https://github.com/planetarium/lib9c/blob/development/.Lib9c.Tests/Action/ExceptionTest.cs#L27-L68).

File: Lib9c/Model/Arena/ArenaScore.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+using Nekoyume.Model.State;
+using Libplanet;
+using Nekoyume.Action;
+
+namespace Nekoyume.Model.Arena
+{
+    /// <summary>
+    /// Introduced at https://github.com/planetarium/lib9c/pull/1006
+    /// </summary>
+    public class ArenaScore : IState
```
[the remark this answers is not in the document]
Comment (hskim881028) [reply Lib9c/Model/Arena/ArenaScore.cs:11]: > I think, ArenaScore & ArenaInformation is managed information for each round. right?, is there any reason to separate class?

The 9c client uses the score to display a list of users. When displaying a list of users, the information in `ArenaInformation` is not needed, so I separated it

File: Lib9c/Model/State/NotExistException.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (hskim881028) [reply Lib9c/Model/State/NotExistException.cs:6]: `NotExistException` is not used. so I delete it

File: Lib9c/Model/State/ArenaInfomation.cs [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+using Bencodex.Types;
+using Libplanet;
+using Nekoyume.Battle;
+using Nekoyume.Model.BattleStatus;
+using Nekoyume.Model.Item;
+using Nekoyume.TableData;
+
+namespace Nekoyume.Model.State
+{
+    public class ArenaInfomation : IState
```
Comment (boscohyun) [inline Lib9c/Model/State/ArenaInfomation.cs:13]: The `Model.State.ArenaInfomation` and the `Model.Arena.ArenaInfomation` look too similar.
This `Model.State.ArenaInfomation` was `Model.State.ArenaInfo` right? If so it is better to use origin name than new name I think.

File: Lib9c/Action/JoinArena.cs [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+            }
+            else
+            {
+                if (roundData.DiscountedEntranceFee > 0)
+                {
+                    return roundData.DiscountedEntranceFee * CrystalCalculator.CRYSTAL;
+                }
+            }
+
+            return 0 * CrystalCalculator.CRYSTAL;
```
[the remark this answers is not in the document]
Comment (hskim881028) [reply Lib9c/Action/JoinArena.cs:195]: Sometimes a fee is 0, so I will reflect it except `throw`

File: Lib9c/Model/Arena/ArenaParticipants.cs [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+            Address = serialized[0].ToAddress();
+            AvatarAddresses = ((List)serialized[1]).Select(c => c.ToAddress()).ToList();
+        }
+
+        public IValue Serialize()
+        {
+            return List.Empty
+                .Add(Address.Serialize())
+                .Add(AvatarAddresses.Aggregate(List.Empty,
+                    (list, address) => list.Add(address.Serialize())));
```
[the remark this answers is not in the document]
Comment (hskim881028) [reply Lib9c/Model/Arena/ArenaParticipants.cs:38]: I wrote it referring to [this](https://github.com/planetarium/lib9c/blob/development/Lib9c/Action/RankingBattle.cs#L274-L276), would it be a problem? 🤔  @boscohyun

**Comment by ipdae** [comment]
test failed

**Pull request closed without merging**